core/trie/bitarray.go

```diff
@@ -0,0 +1,397 @@
+package trie
+
+import (
+	"bytes"
+	"encoding/binary"
+	"math"
+	"math/bits"
+
+	"github.com/NethermindEth/juno/core/felt"
+)
+
+const (
+	maxUint64 = uint64(math.MaxUint64)
+	bits8     = 8
+)
+
+var (
+	maxBitArray   = [4]uint64{math.MaxUint64, math.MaxUint64, math.MaxUint64, math.MaxUint64}
+	emptyBitArray = new(BitArray)
+)
+
+// BitArray is a structure that represents a bit array with length representing the number of used bits.
+// It uses a little endian representation to do bitwise operations of the words efficiently.
+// For example, if len is 10, it means that the 2^9, 2^8, ..., 2^0 bits are used.
+// The max length is 255 bits (uint8), because our use case only need up to 251 bits for a given trie key.
+// Although words can be used to represent 256 bits, we don't want to add an additional byte for the length.
+type BitArray struct {
+	len   uint8     // number of used bits
+	words [4]uint64 // little endian (i.e. words[0] is the least significant)
+}
+
+func (b *BitArray) Felt() *felt.Felt {
+	bs := b.Bytes()
+	return new(felt.Felt).SetBytes(bs[:])
+}
+
+func (b *BitArray) Len() uint8 {
+	return b.len
+}
+
+// Bytes returns the bytes representation of the bit array in big endian format
+//
+//nolint:mnd
+func (b *BitArray) Bytes() [32]byte {
+	var res [32]byte
+
+	switch {
+	case b.len == 0:
+		// all zeros
+		return res
+	case b.len >= 192:
+		// Create mask for top word: keeps only valid bits above 192
+		// e.g., if len=200, keeps lowest 8 bits (200-192)
+		mask := maxUint64 >> (256 - uint16(b.len))
+		binary.BigEndian.PutUint64(res[0:8], b.words[3]&mask)
+		binary.BigEndian.PutUint64(res[8:16], b.words[2])
+		binary.BigEndian.PutUint64(res[16:24], b.words[1])
+		binary.BigEndian.PutUint64(res[24:32], b.words[0])
+	case b.len >= 128:
+		// Mask for bits 128-191: keeps only valid bits above 128
+		// e.g., if len=150, keeps lowest 22 bits (150-128)
+		mask := maxUint64 >> (192 - b.len)
+		binary.BigEndian.PutUint64(res[8:16], b.words[2]&mask)
+		binary.BigEndian.PutUint64(res[16:24], b.words[1])
+		binary.BigEndian.PutUint64(res[24:32], b.words[0])
+	case b.len >= 64:
+		// You get the idea
+		mask := maxUint64 >> (128 - b.len)
+		binary.BigEndian.PutUint64(res[16:24], b.words[1]&mask)
+		binary.BigEndian.PutUint64(res[24:32], b.words[0])
+	default:
+		mask := maxUint64 >> (64 - b.len)
+		binary.BigEndian.PutUint64(res[24:32], b.words[0]&mask)
+	}
+
+	return res
+}
+
+// EqualMSBs checks if two bit arrays share the same most significant bits, where the length of
+// the check is determined by the shorter array. Returns true if either array has
+// length 0, or if the first min(b.len, x.len) MSBs are identical.
```

Comment:
Since it is written from the point of view of a method I would change the description to represent it. Something along the lines of:

```
// Checks if the current bit array shares the same most significant bits with
// another bit array ...
```

Just suggestions though, feel free dismiss any part of it